# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
load_dotenv()
%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\ugokh\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!

Bad key savefig.frameon in file c:\Users\ugokh\anaconda3\envs\pyvizenv\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 421 ('savefig.frameon : True')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.4.3/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.level in file c:\Users\ugokh\anaconda3\envs\pyvizenv\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 472 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.4.3/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.fileo in file c:\Users\ugokh\anaconda3\envs\pyvizen

In [2]:
# Read your api key environment variable
apiKey = os.getenv("NEWS_API_KEY")

In [3]:
# Create a newsapi client
from newsapi import NewsApiClient
newsApi = NewsApiClient(api_key=apiKey)

In [4]:
# Fetch the Bitcoin news articles
bitcoinArticles = newsApi.get_everything(q="bitcoin", language="en", sort_by="relevancy")

In [5]:
# Fetch the Ethereum news articles
ethereumArticles = newsApi.get_everything(q="ethereum", language="en", sort_by="relevancy")

In [7]:
# Create the Bitcoin sentiment scores DataFrame
btcSentimentScore = []

for article in bitcoinArticles["articles"]:
    try:
        text = article["content"]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btcSentimentScore.append({
            "text": text,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
        })
    except ArithmeticError:
        pass
    
#btcSentimentScore

In [ ]:
#creating the btc dataframe so that results looks clean and understandable (do the same for eth)
btcDf = pd.DataFrame(btcSentimentScore)
colReorder = ["compound", "negative", "neutral", "positive", "text"]
btcDf = btcDf[colReorder]

In [8]:
btcDf

compound  negative  neutral  positive  \
0     0.0772     0.000    0.964     0.036   
1    -0.0516     0.061    0.882     0.056   
2    -0.3269     0.143    0.772     0.085   
3    -0.1027     0.040    0.960     0.000   
4     0.3400     0.000    0.928     0.072   
5     0.3818     0.052    0.833     0.114   
6     0.3182     0.040    0.883     0.077   
7     0.7506     0.000    0.807     0.193   
8    -0.4404     0.241    0.557     0.202   
9    -0.4767     0.103    0.897     0.000   
10   -0.3612     0.123    0.877     0.000   
11   -0.3818     0.077    0.923     0.000   
12   -0.2732     0.063    0.937     0.000   
13   -0.3818     0.077    0.923     0.000   
14    0.1027     0.066    0.859     0.076   
15   -0.5499     0.106    0.894     0.000   
16    0.1326     0.000    0.956     0.044   
17    0.2263     0.060    0.855     0.085   
18   -0.4019     0.083    0.879     0.039   
19   -0.2960     0.084    0.838     0.079   

                                                 text  
0   Now, even though there are a number of women-f...  
1   A Bitcoin mining site powered by otherwise los...  
2   Warren Buffett has always been a bitcoin skept...  
3   You can now reportedly pay for your burritos a...  
4   Image source, Getty Images\r\nThe value of Bit...  
5   As a kid, I remember when my father tried to u...  
6   Customers at Chipotle will now be able to pay ...  
7   If youve ever felt like introducing some Vegas...  
8   Cryptocurrency mixers are sometimes used to he...  
9   Photo Illustration by Grayson Blackmon / The V...  
10  Posted \r\nFrom Bitcoin's dramatic drop to a n...  
11  May 11 (Reuters) - Bitcoin fell 7.23% to $28,7...  
12  May 4 (Reuters) - Bitcoin rose 5.7% to $39,862...  
13  May 9 (Reuters) - Bitcoin dropped 7.81% to $31...  
14  BTC fell to its lowest level since last July, ...  
15  At a Mexican restaurant in North London a few ...  
16  May 24 (Reuters) - The crypto winter is into i...  
17  May 10 (Reuters) - Bitcoin has scant experienc...  
18  Terra is launching a new blockchain with a new...  
19  Terra launched a new blockchain with a new lun...

In [ ]:
# Create the Ethereum sentiment scores DataFrame
ethSentimentScore = []

for article in ethereumArticles["articles"]:
    try:
        text = article["content"]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        ethSentimentScore.append({
            "text": text,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
        })
    except ArithmeticError:
        pass
    
ethSentimentScore

In [12]:
#create eth df and reorganize the columns
ethDf = pd.DataFrame(ethSentimentScore)

ethDf = ethDf[colReorder]

ethDf

compound  negative  neutral  positive  \
0     0.6486     0.000    0.865     0.135   
1    -0.1027     0.040    0.960     0.000   
2    -0.2263     0.075    0.879     0.046   
3    -0.2732     0.055    0.945     0.000   
4     0.1280     0.000    0.954     0.046   
5     0.0258     0.000    0.966     0.034   
6     0.6908     0.000    0.822     0.178   
7    -0.6908     0.178    0.822     0.000   
8    -0.3818     0.085    0.847     0.069   
9    -0.2732     0.063    0.937     0.000   
10   -0.3818     0.077    0.923     0.000   
11   -0.3818     0.077    0.923     0.000   
12    0.0000     0.000    1.000     0.000   
13   -0.1027     0.036    0.964     0.000   
14    0.1027     0.066    0.859     0.076   
15   -0.1531     0.045    0.955     0.000   
16   -0.3182     0.067    0.933     0.000   
17    0.3400     0.000    0.928     0.072   
18   -0.4767     0.100    0.900     0.000   
19   -0.1531     0.045    0.955     0.000   

                                                 text  
0   Meta has revealed more of how NFTs will work o...  
1   GameStop has officially thrown itself headlong...  
2   A multi-billion dollar cryptocurrency company ...  
3   When Bored Ape Yacht Club creators Yuga Labs a...  
4   GameStop is going all-in on crypto. The video ...  
5   DAVOS, Switzerland, May 25 (Reuters) - Ethereu...  
6   Editorial IndependenceWe want to help you make...  
7   40 days ago Bitcoin sold for $47,454. It's pri...  
8   When Nvidia launched its Ampere Lite Hash Rate...  
9   May 4 (Reuters) - Bitcoin rose 5.7% to $39,862...  
10  May 11 (Reuters) - Bitcoin fell 7.23% to $28,7...  
11  May 9 (Reuters) - Bitcoin dropped 7.81% to $31...  
12  9 with 8 posters participating\r\nNvidia began...  
13  If you ever wanted to buy an NFT based on Ethe...  
14  BTC fell to its lowest level since last July, ...  
15  The cryptocurrency boom over the past few year...  
16  ApeCoin, the native cryptocurrency of the APE ...  
17  Image source, Getty Images\r\nThe value of Bit...  
18  Solana, the third largest cryptocurrency by ma...  
19  The cryptocurrency boom over the past few year...

In [13]:
# Describe the Bitcoin Sentiment
btcDf.describe()

compound   negative    neutral   positive
count  20.000000  20.000000  20.000000  20.000000
mean   -0.085735   0.070950   0.871200   0.057900
std     0.359368   0.057206   0.089972   0.060467
min    -0.549900   0.000000   0.557000   0.000000
25%    -0.381800   0.040000   0.850750   0.000000
50%    -0.187950   0.064500   0.882500   0.050000
75%     0.156025   0.088750   0.924250   0.080500
max     0.750600   0.241000   0.964000   0.202000

In [14]:
# Describe the Ethereum Sentiment
ethDf.describe()

compound   negative    neutral   positive
count  20.000000  20.000000  20.000000  20.000000
mean   -0.098975   0.050450   0.916850   0.032800
std     0.352184   0.044882   0.051293   0.051042
min    -0.690800   0.000000   0.822000   0.000000
25%    -0.334100   0.000000   0.875500   0.000000
50%    -0.153100   0.050000   0.930500   0.000000
75%     0.045025   0.075500   0.955000   0.051750
max     0.690800   0.178000   1.000000   0.178000

### Questions:

Q: Which coin had the highest mean positive score?

A: Btc with 0.058

Q: Which coin had the highest compound score?

A: Btc with 0.750

Q. Which coin had the highest positive score?

A: Btc with 0.202

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [10]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [11]:
# Instantiate the lemmatizer
# YOUR CODE HERE!

# Create a list of stopwords
# YOUR CODE HERE!

# Expand the default stopwords list if necessary
# YOUR CODE HERE!

In [12]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text

   
    # Create a tokenized list of the words
    
    
    # Lemmatize words into root words

   
    # Convert the words to lowercase
    
    
    # Remove the stop words
    
    
    return tokens

In [13]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!

In [14]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---